### ISSUE TITLE
AIX and DL_PrivateKey_GFP<GP>::~DL_PrivateKey_GFP() causes a section type conflict

### ISSUE THREAD

**Issue by noloader** [weidai11/cryptopp#499]
Well, this is one of the more bizarre errors I've seen from GCC. Its GCC 6.1.0 on AIX 7.2.0.0, which is provided by GCC119 on the compile farm.

```
$ gmake -j 64
...
g++ -DNDEBUG -g2 -O3 -fPIC -pthread -pipe -c bench1.cpp
g++ -DNDEBUG -g2 -O3 -fPIC -pthread -pipe -c bench2.cpp
...
In file included from bench2.cpp:9:0:
gfpcrypt.h: In function 'CryptoPP::DL_PrivateKey_GFP<GP>::~DL_PrivateKey_GFP() [with GP = CryptoPP::DL_GroupParameters_DSA]':
gfpcrypt.h:519:13: error: CryptoPP::DL_PrivateKey_GFP<GP>::~DL_PrivateKey_GFP() [with GP = CryptoPP::DL_GroupParameters_DSA] causes a section type conflict with void CryptoPP::AllocatorWithCleanup<T, T_Align16>::deallocate(void*, CryptoPP::AllocatorWithCleanup<T, T_Align16>::size_type) [with T = unsigned char; bool T_Align16 = true]
     virtual ~DL_PrivateKey_GFP() {}
             ^
In file included from integer.h:20:0,
                 from validate.h:8,
                 from bench2.cpp:6:
secblock.h:217:7: note: 'void CryptoPP::AllocatorWithCleanup<T, T_Align16>::deallocate(void*, CryptoPP::AllocatorWithCleanup<T, T_Align16>::size_type) [with T = unsigned char; bool T_Align16 = true]' was declared here
  void deallocate(void *ptr, size_type size)
       ^~~~~~~~~~
GNUmakefile:1026: recipe for target 'bench2.o' failed
```

**Comment by noloader** [comment]
We chased this issue for several days, including [How to get more information on “X causes a section type conflict with Y”](https://stackoverflow.com/q/46211621/608639) on Stack Overflow.

Here is the fix committed at [51752cb91aa6](https://github.com/weidai11/cryptopp/commit/51752cb91aa6). We have no idea what caused the issue or why it fixed it.

```
$ git diff
diff --git a/gfpcrypt.h b/gfpcrypt.h
index ae05b48..98047c6 100644
--- a/gfpcrypt.h
+++ b/gfpcrypt.h
@@ -516,7 +516,7 @@ template <class GP>
 class DL_PrivateKey_GFP : public DL_PrivateKeyImpl<GP>
 {
 public:
-    virtual ~DL_PrivateKey_GFP() {}
+    virtual ~DL_PrivateKey_GFP();^M

     //! \brief Create a private key
     //! \param rng a RandomNumberGenerator derived class
@@ -570,6 +570,9 @@ public:
         {this->AccessGroupParameters().Initialize(p, q, g); this->SetPrivateExponent(x);}
 };

+template <class GP>^M
+DL_PrivateKey_GFP<GP>::~DL_PrivateKey_GFP() {}^M
+^M
 //! \class DL_SignatureKeys_GFP
 //! \brief Discrete Log (DL) signing/verification keys in GF(p) groups
 struct DL_SignatureKeys_GFP
```

**Comment by noloader** [comment]
We chased this issue for several days, including [How to get more information on “X causes a section type conflict with Y”](https://stackoverflow.com/q/46211621/608639) on Stack Overflow.

Here is the fix committed at [51752cb91aa6](https://github.com/weidai11/cryptopp/commit/51752cb91aa6). We have no idea what caused the issue or why it fixed it.

```
$ git diff
diff --git a/gfpcrypt.h b/gfpcrypt.h
index ae05b48..c3260ca 100644
--- a/gfpcrypt.h
+++ b/gfpcrypt.h
@@ -516,7 +516,7 @@ template <class GP>
 class DL_PrivateKey_GFP : public DL_PrivateKeyImpl<GP>
 {
 public:
-    virtual ~DL_PrivateKey_GFP() {}
+    virtual ~DL_PrivateKey_GFP();^M

     //! \brief Create a private key
     //! \param rng a RandomNumberGenerator derived class
@@ -570,6 +570,10 @@ public:
         {this->AccessGroupParameters().Initialize(p, q, g); this->SetPrivateExponent(x);}
 };

+// Out-of-line dtor due to AIX and GCC, http://github.com/weidai11/cryptopp/issues/499^M
+template <class GP>^M
+DL_PrivateKey_GFP<GP>::~DL_PrivateKey_GFP() {}^M
+^M
 //! \class DL_SignatureKeys_GFP
 //! \brief Discrete Log (DL) signing/verification keys in GF(p) groups
 struct DL_SignatureKeys_GFP
```
